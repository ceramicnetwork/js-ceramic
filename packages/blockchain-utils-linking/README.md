# Utils for linking blockchain accounts

> This package contains utility functions to link blockchain accounts to a DID.

### Installation
```shell
npm install @ceramicnetwork/blockchain-utils-linking
```

## Contributing
We are happy to accept small and large contributions. Make sure to check out the [Ceramic specifications](https://github.com/ceramicnetwork/specs) for details of how the protocol works.

## License

Apache-2.0 or MIT
