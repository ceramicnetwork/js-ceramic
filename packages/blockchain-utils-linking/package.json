{
  "name": "@ceramicnetwork/blockchain-utils-linking",
  "version": "0.0.6-rc.0",
  "description": "Blockchain utils for linking blockchain accounts to DID",
  "keywords": [
    "ceramic",
    "caip10-link"
  ],
  "homepage": "https://ceramic.network",
  "license": "(Apache-2.0 OR MIT)",
  "directories": {
    "lib": "lib"
  },
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "test": "../../node_modules/.bin/jest",
    "build": "../../node_modules/.bin/tsc -p tsconfig.json",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean",
    "lint": "../../node_modules/.bin/eslint ./src --ext .js,.jsx,.ts,.tsx",
    "clean": "rm -rf ./lib"
  },
  "dependencies": {
    "@stablelib/sha256": "^1.0.0",
    "caip": "^0.9.2",
    "uint8arrays": "^2.0.5"
  },
  "devDependencies": {
    "@glif/filecoin-address": "^1.1.0-beta.15",
    "@glif/local-managed-provider": "^1.1.0-beta.15",
    "@polkadot/api": "^3.0.1",
    "@polkadot/keyring": "^5.0.1",
    "@polkadot/types": "^3.0.1",
    "@polkadot/util": "^5.0.1",
    "@polkadot/util-crypto": "^5.2.3",
    "@smontero/eosio-local-provider": "^0.0.3",
    "@tendermint/sig": "^0.6.0",
    "@zondax/filecoin-signing-tools": "^0.13.0",
    "eth-sig-util": "^3.0.0",
    "ganache-core": "^2.13.1",
    "near-api-js": "^0.36.3"
  },
  "jest": {
    "testEnvironment": "jest-environment-uint8array",
    "transformIgnorePatterns": [
      "/node_modules(?!/did-jwt)/"
    ]
  },
  "gitHead": "34eeee25597b0a60def72906c26d3afd6230aaf1"
}
