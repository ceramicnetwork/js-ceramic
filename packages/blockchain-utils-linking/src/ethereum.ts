import { AuthProvider } from "./auth-provider";
import { AccountID } from "caip";
import {
  encodeRpcMessage,
  getConsentMessage,
  LinkProof,
  RpcMessage,
} from "./util";
import * as uint8arrays from "uint8arrays";
import * as sha256 from '@stablelib/sha256'

const ADDRESS_TYPES = {
  ethereumEOA: "ethereum-eoa",
  erc1271: "erc1271",
};

const CHAIN_NAMESPACE = "eip155";

/**
 *  AuthProvider which can be used for Ethereum providers with standard interface
 */
export class EthereumAuthProvider implements AuthProvider {
  readonly isAuthProvider = true;

  constructor(
      private readonly provider: any,
      private readonly address: string
  ) {}

  async accountId() {
    const payload = encodeRpcMessage("eth_chainId", []);
    const chainIdHex = await safeSend(payload, this.provider);
    const chainId = parseInt(chainIdHex, 16);
    return new AccountID({
      address: this.address,
      chainId: `${CHAIN_NAMESPACE}:${chainId}`,
    });
  }

  async authenticate(message: string): Promise<string> {
    const accountId = await this.accountId();
    return authenticate(message, accountId, this.provider);
  }

  async createLink(did: string): Promise<LinkProof> {
    const accountId = await this.accountId();
    return createLink(did, accountId, this.provider, { type: "ethereum-eoa" });
  }

  withAddress(address: string): AuthProvider {
    return new EthereumAuthProvider(this.provider, address);
  }
}

export function isEthAddress(address: string): boolean {
  return /^0x[a-fA-F0-9]{40}$/.test(address);
}

async function getCode(address: string, provider: any): Promise<string> {
  const payload = encodeRpcMessage("eth_getCode", [address, "latest"]);
  return safeSend(payload, provider);
}

async function safeSend(data: RpcMessage, provider: any): Promise<any> {
  const send = (provider.sendAsync ? provider.sendAsync : provider.send).bind(
      provider
  );
  return new Promise((resolve, reject) => {
    send(data, function (err: any, result: any) {
      if (err) reject(err);
      else if (result.error) reject(result.error);
      else resolve(result.result);
    });
  });
}

export async function isERC1271(
    account: AccountID,
    provider: any
): Promise<boolean> {
  const bytecode = await getCode(account.address, provider).catch(() => null);
  return Boolean(
      bytecode && bytecode !== "0x" && bytecode !== "0x0" && bytecode !== "0x00"
  );
}

export function normalizeAccountId(account: AccountID): AccountID {
  account.address = account.address.toLowerCase();
  return account;
}

function utf8toHex(message: string): string {
  const bytes = uint8arrays.fromString(message);
  const hex = uint8arrays.toString(bytes, "base16");
  return "0x" + hex;
}

async function createEthLink(
    did: string,
    account: AccountID,
    provider: any,
    opts: any = {}
): Promise<LinkProof> {
  const { message, timestamp } = getConsentMessage(did, !opts.skipTimestamp);
  const hexMessage = utf8toHex(message);
  const payload = encodeRpcMessage("personal_sign", [
    hexMessage,
    account.address,
  ]);
  const signature = await safeSend(payload, provider);
  const proof: LinkProof = {
    version: 2,
    type: ADDRESS_TYPES.ethereumEOA,
    message,
    signature,
    account: account.toString(),
  };
  if (!opts.skipTimestamp) proof.timestamp = timestamp;
  return proof;
}

async function validateChainId(
    account: AccountID,
    provider: any
): Promise<void> {
  const payload = encodeRpcMessage("eth_chainId", []);
  const chainIdHex = await safeSend(payload, provider);
  const chainId = parseInt(chainIdHex, 16);
  if (chainId !== parseInt(account.chainId.reference)) {
    throw new Error(
        `ChainId in provider (${chainId}) is different from AccountID (${account.chainId.reference})`
    );
  }
}

async function createErc1271Link(
    did: string,
    account: AccountID,
    provider: any,
    opts: any
): Promise<LinkProof> {
  const ethLinkAccount = opts?.eoaSignAccount || account;
  const res = await createEthLink(did, ethLinkAccount, provider, opts);
  await validateChainId(account, provider);
  return Object.assign(res, {
    type: ADDRESS_TYPES.erc1271,
    account: account.toString(),
  });
}

export async function createLink(
    did: string,
    account: AccountID,
    provider: any,
    opts: any
): Promise<LinkProof> {
  account = normalizeAccountId(account);
  if (await isERC1271(account, provider)) {
    return createErc1271Link(did, account, provider, opts);
  } else {
    return createEthLink(did, account, provider, opts);
  }
}

export async function authenticate(
    message: string,
    account: AccountID,
    provider: any
): Promise<string> {
  if (account) account = normalizeAccountId(account);
  if (provider.isAuthereum) return provider.signMessageWithSigningKey(message);
  const hexMessage = utf8toHex(message);
  const payload = encodeRpcMessage("personal_sign", [
    hexMessage,
    account.address,
  ]);
  const signature = await safeSend(payload, provider);
  const signatureBytes = uint8arrays.fromString(signature.slice(2))
  const digest = sha256.hash(signatureBytes)
  return `0x${uint8arrays.toString(digest, 'base16')}`;
}
