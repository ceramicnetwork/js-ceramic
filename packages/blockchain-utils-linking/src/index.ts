export * from "./auth-provider";
export * as ethereum from "./ethereum";
export * from "./util";
export * as filecoin from "./filecoin";
export * as polkadot from "./polkadot";
export * as eosio from "./eosio";
