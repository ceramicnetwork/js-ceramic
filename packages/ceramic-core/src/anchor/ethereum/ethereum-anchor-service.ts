import CID from "cids";
import fetch from "cross-fetch";

import { decode } from "typestub-multihashes";
import * as providers from "@ethersproject/providers"
import { CeramicConfig } from "../../ceramic";

import AnchorServiceResponse from "../anchor-service-response";
import { AnchorService } from "@ceramicnetwork/ceramic-common"
import { AnchorProof } from "@ceramicnetwork/ceramic-common"

/**
 * CID-docId pair
 */
interface CidDoc {
    readonly cid: CID;
    readonly docId: string;
}

const DEFAULT_POLL_TIME = 5000; // 5 seconds
const DEFAULT_MAX_POLL_TIME = 7200000; // 2 hours

/**
 * Ethereum network configuration
 */
interface EthNetwork {
    network: string;
    chain: string;
    chainId: number;
    networkId: number;
    type: string;
}

/**
 * Maps some of Ethereum chain IDs to network configuration
 */
const ETH_CHAIN_ID_MAPPINGS: Record<string, EthNetwork> = {
    "eip155:1": { network: "mainnet", chain: "ETH", chainId: 1, networkId: 1, type: "Production" },
    "eip155:3": { network: "ropsten", chain: "ETH", chainId: 3, networkId: 3, type: "Test" },
};

/**
 * Ethereum anchor service that stores root CIDs on Ethereum blockchain
 */
export default class EthereumAnchorService extends AnchorService {

    private readonly cidToResMap: Map<CidDoc, AnchorServiceResponse>;

    /**
     * @param _config - service configuration (polling interval, etc.)
     */
    constructor(private _config: CeramicConfig) {
        super();

        this.cidToResMap = new Map<CidDoc, AnchorServiceResponse>();
    }

    /**
     * Requests anchoring service for current head of the document
     * @param docId - Document ID
     * @param head - Head CID of the document
     */
    async requestAnchor(docId: string, head: CID): Promise<void> {
        const cidDocPair: CidDoc = { cid: head, docId };

        // send initial request
        await this._sendReq(cidDocPair);
        this._poll(cidDocPair); // start polling
    }

    /**
     * Send requests to an external Ceramic Anchor Service
     * @param cidDocPair - mapping
     * @private
     */
    async _sendReq(cidDocPair: CidDoc): Promise<void> {
        const response = await fetch(this._config.anchorServiceUrl, {
            method: "POST", body: JSON.stringify({
                docId: cidDocPair.docId, cid: cidDocPair.cid.toString()
            }), headers: {
                "Content-Type": "application/json"
            }
        });

        if (!response.ok) {
            this.emit(cidDocPair.docId, {
                status: 'FAILED',
                message: `Failed to send request. Status ${response.statusText}`
            });
            return;
        }
        const json = await response.json();

        const res = { status: json.status, message: json.message, anchorScheduledFor: json.scheduledAt };
        this.cidToResMap.set(cidDocPair, res);
        this.emit(cidDocPair.docId, res);
    }

    /**
     * Start polling for CidDocPair mapping
     * @param cidDoc - CID to Doc mapping
     * @param pollTime - Single poll timeout
     * @param maxPollingTime - Global timeout for max polling in milliseconds
     * @private
     */
    async _poll(cidDoc: CidDoc, pollTime?: number, maxPollingTime?: number): Promise<void> {
        const started = new Date().getTime();
        const maxTime = started + (maxPollingTime | DEFAULT_MAX_POLL_TIME);

        let poll = true;
        while (poll) {
            if (started > maxTime) {
                const failedRes = { status: 'FAILED', message: 'exceeded max timeout' };
                this.cidToResMap.set(cidDoc, failedRes);

                this.emit(cidDoc.docId, failedRes);
                return; // exit loop
            }

            await new Promise(resolve => setTimeout(resolve, DEFAULT_POLL_TIME));

            try {
                const requestUrl = [this._config.anchorServiceUrl, cidDoc.cid.toString()].join('/');
                const response = await fetch(requestUrl);
                const json = await response.json();

                switch (json.status) {
                    case "PENDING": {
                        // just log
                        break;
                    }
                    case "PROCESSING": {
                        this.emit(cidDoc.docId, { status: json.status, message: json.message });
                        break;
                    }
                    case "FAILED": {
                        this.emit(cidDoc.docId, { status: json.status, message: json.message });
                        poll = false;
                        break;
                    }
                    case "COMPLETED": {
                        const { anchorRecord } = json;
                        const anchorRecordCid = new CID(anchorRecord.cid.toString());

                        this.emit(cidDoc.docId, {
                            status: json.status, message: json.message, anchorRecord: anchorRecordCid
                        });
                        poll = false;
                        break;
                    }
                }
            } catch (e) {
                // just log
            }
        }
    }

    /**
     * Validate anchor proof on the chain
     * @param anchorProof - Anchor proof instance
     */
    async validateChainInclusion(anchorProof: AnchorProof): Promise<void> {
        const decoded = decode(anchorProof.txHash.multihash);
        const txHash = decoded.digest.toString("hex");

        // determine network based on a chain ID
        const provider: providers.BaseProvider = this._getEthProvider(anchorProof.chainId);

        const transaction = await provider.getTransaction('0x' + txHash);
        const block = await provider.getBlock(transaction.blockHash);

        const txValueHexNumber = parseInt(transaction.data, 16);
        const rootValueHexNumber = parseInt('0x' + anchorProof.root.toBaseEncodedString('base16'), 16);

        if (txValueHexNumber !== rootValueHexNumber) {
            throw new Error(`The root CID ${anchorProof.root.toString()} is not in the transaction`);
        }

        if (anchorProof.blockNumber !== transaction.blockNumber) {
            throw new Error(`Block numbers are not the same`);
        }

        if (anchorProof.blockTimestamp !== block.timestamp) {
            throw new Error(`Block timestamps are not the same`);
        }
    }

    /**
     * Gets Ethereum provider based on chain ID
     * @param chain - CAIP-2 Chain ID
     * @private
     */
    private _getEthProvider(chain: string): providers.BaseProvider {
        if (!chain.startsWith('eip155')) {
            throw new Error('Invalid chain ID according to CAIP-2');
        }

        const ethNetwork: EthNetwork = ETH_CHAIN_ID_MAPPINGS[chain];
        if (ethNetwork == null) {
            // defaults to configuration Ethereum RPC URL
            return new providers.JsonRpcProvider(this._config.ethereumRpcUrl);
        }

        return providers.getDefaultProvider(ethNetwork.network);
    }

}
