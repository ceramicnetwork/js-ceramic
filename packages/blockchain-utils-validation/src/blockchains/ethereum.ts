import { verifyMessage } from "@ethersproject/wallet";
import { Contract } from "@ethersproject/contracts";
import * as providers from "@ethersproject/providers";
import { AccountID } from "caip";
import * as uint8arrays from "uint8arrays";
import { LinkProof } from "@ceramicnetwork/blockchain-utils-linking";
import { BlockchainHandler } from "../blockchain-handler";

const ADDRESS_TYPES = {
  ethereumEOA: "ethereum-eoa",
  erc1271: "erc1271",
};
const ERC1271_ABI = [
  "function isValidSignature(bytes _messageHash, bytes _signature) public view returns (bytes4 magicValue)",
];
const MAGIC_ERC1271_VALUE = "0x20c13b0b";
const namespace = "eip155";

function utf8toHex(message: string): string {
  const bytes = uint8arrays.fromString(message);
  const hex = uint8arrays.toString(bytes, "base16");
  return "0x" + hex;
}

function getEthersProvider(chainId: string): any {
  const network = providers.getNetwork(chainId);
  if (!network._defaultProvider)
    throw new Error(`Network with chainId ${chainId} is not supported`);
  return network._defaultProvider(providers);
}

function toV2Proof(proof: LinkProof, address?: string): LinkProof {
  proof.account = new AccountID({
    address: (proof.version === 1 ? proof.address : address) || "",
    chainId: {
      namespace,
      reference: proof.chainId ? proof.chainId.toString() : "1",
    },
  }).toString();
  delete proof.address;
  delete proof.chainId;
  proof.version = 2;
  return proof;
}

async function validateEoaLink(proof: LinkProof): Promise<LinkProof | null> {
  const recoveredAddr = verifyMessage(
    proof.message,
    proof.signature
  ).toLowerCase();
  if (proof.version !== 2) proof = toV2Proof(proof, recoveredAddr);
  const account = new AccountID(proof.account);
  if (account.address !== recoveredAddr) {
    return null;
  }
  return proof;
}

async function validateErc1271Link(
  proof: LinkProof
): Promise<LinkProof | null> {
  if (proof.version === 1) proof = toV2Proof(proof);
  const account = new AccountID(proof.account);
  const provider = getEthersProvider(account.chainId.reference);
  const contract = new Contract(account.address, ERC1271_ABI, provider);
  const message = utf8toHex(proof.message);
  const returnValue = await contract.isValidSignature(message, proof.signature);

  return returnValue === MAGIC_ERC1271_VALUE ? proof : null;
}

async function validateLink(proof: LinkProof): Promise<LinkProof | null> {
  try {
    if (proof.type === ADDRESS_TYPES.erc1271) {
      return await validateErc1271Link(proof);
    } else {
      return await validateEoaLink(proof);
    }
  } catch {
    return null;
  }
}

const handler: BlockchainHandler = {
  namespace,
  validateLink,
};

export default handler;
