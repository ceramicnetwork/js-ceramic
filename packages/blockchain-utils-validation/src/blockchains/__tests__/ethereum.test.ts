import { AccountID } from "caip";
import ganache from "ganache-core";
import ethereum from "../ethereum";
import * as sigUtils from "eth-sig-util";
import { ContractFactory, Contract } from "@ethersproject/contracts";
import * as providers from "@ethersproject/providers";
import * as linking from "@ceramicnetwork/blockchain-utils-linking";
import { proofs } from "./fixtures";
import {LinkProof} from "@ceramicnetwork/blockchain-utils-linking";

const CONTRACT_WALLET_ABI = [
  {
    constant: false,
    inputs: [{ internalType: "bool", name: "valid", type: "bool" }],
    name: "setIsValid",
    outputs: [],
    payable: false,
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    constant: true,
    inputs: [
      { internalType: "bytes", name: "_messageHash", type: "bytes" },
      { internalType: "bytes", name: "_signature", type: "bytes" },
    ],
    name: "isValidSignature",
    outputs: [{ internalType: "bytes4", name: "magicValue", type: "bytes4" }],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
];
const CONTRACT_WALLET_BYTECODE = {
  linkReferences: {},
  object:
    "608060405260008060006101000a81548160ff02191690831515021790555034801561002a57600080fd5b506102938061003a6000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c806317c136771461003b57806320c13b0b1461006b575b600080fd5b6100696004803603602081101561005157600080fd5b8101908080351515906020019092919050505061020f565b005b6101bb6004803603604081101561008157600080fd5b810190808035906020019064010000000081111561009e57600080fd5b8201836020820111156100b057600080fd5b803590602001918460018302840111640100000000831117156100d257600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f8201169050808301925050505050505091929192908035906020019064010000000081111561013557600080fd5b82018360208201111561014757600080fd5b8035906020019184600183028401116401000000008311171561016957600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f82011690508083019250505050505050919291929050505061022b565b60405180827bffffffffffffffffffffffffffffffffffffffffffffffffffffffff19167bffffffffffffffffffffffffffffffffffffffffffffffffffffffff1916815260200191505060405180910390f35b806000806101000a81548160ff02191690831515021790555050565b60008060009054906101000a900460ff1615610250576320c13b0b60e01b9050610258565b600060e01b90505b9291505056fea265627a7a723158209d7aa06b7443aa12cee8b1ba4356af624de6b912d6af47b494c9b9d621b883ac64736f6c634300050b0032",
  opcodes:
    "PUSH1 0x80 PUSH1 0x40 MSTORE PUSH1 0x0 DUP1 PUSH1 0x0 PUSH2 0x100 EXP DUP2 SLOAD DUP2 PUSH1 0xFF MUL NOT AND SWAP1 DUP4 ISZERO ISZERO MUL OR SWAP1 SSTORE POP CALLVALUE DUP1 ISZERO PUSH2 0x2A JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH2 0x293 DUP1 PUSH2 0x3A PUSH1 0x0 CODECOPY PUSH1 0x0 RETURN INVALID PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH2 0x10 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0x4 CALLDATASIZE LT PUSH2 0x36 JUMPI PUSH1 0x0 CALLDATALOAD PUSH1 0xE0 SHR DUP1 PUSH4 0x17C13677 EQ PUSH2 0x3B JUMPI DUP1 PUSH4 0x20C13B0B EQ PUSH2 0x6B JUMPI JUMPDEST PUSH1 0x0 DUP1 REVERT JUMPDEST PUSH2 0x69 PUSH1 0x4 DUP1 CALLDATASIZE SUB PUSH1 0x20 DUP2 LT ISZERO PUSH2 0x51 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST DUP2 ADD SWAP1 DUP1 DUP1 CALLDATALOAD ISZERO ISZERO SWAP1 PUSH1 0x20 ADD SWAP1 SWAP3 SWAP2 SWAP1 POP POP POP PUSH2 0x20F JUMP JUMPDEST STOP JUMPDEST PUSH2 0x1BB PUSH1 0x4 DUP1 CALLDATASIZE SUB PUSH1 0x40 DUP2 LT ISZERO PUSH2 0x81 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST DUP2 ADD SWAP1 DUP1 DUP1 CALLDATALOAD SWAP1 PUSH1 0x20 ADD SWAP1 PUSH5 0x100000000 DUP2 GT ISZERO PUSH2 0x9E JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST DUP3 ADD DUP4 PUSH1 0x20 DUP3 ADD GT ISZERO PUSH2 0xB0 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST DUP1 CALLDATALOAD SWAP1 PUSH1 0x20 ADD SWAP2 DUP5 PUSH1 0x1 DUP4 MUL DUP5 ADD GT PUSH5 0x100000000 DUP4 GT OR ISZERO PUSH2 0xD2 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST SWAP2 SWAP1 DUP1 DUP1 PUSH1 0x1F ADD PUSH1 0x20 DUP1 SWAP2 DIV MUL PUSH1 0x20 ADD PUSH1 0x40 MLOAD SWAP1 DUP2 ADD PUSH1 0x40 MSTORE DUP1 SWAP4 SWAP3 SWAP2 SWAP1 DUP2 DUP2 MSTORE PUSH1 0x20 ADD DUP4 DUP4 DUP1 DUP3 DUP5 CALLDATACOPY PUSH1 0x0 DUP2 DUP5 ADD MSTORE PUSH1 0x1F NOT PUSH1 0x1F DUP3 ADD AND SWAP1 POP DUP1 DUP4 ADD SWAP3 POP POP POP POP POP POP POP SWAP2 SWAP3 SWAP2 SWAP3 SWAP1 DUP1 CALLDATALOAD SWAP1 PUSH1 0x20 ADD SWAP1 PUSH5 0x100000000 DUP2 GT ISZERO PUSH2 0x135 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST DUP3 ADD DUP4 PUSH1 0x20 DUP3 ADD GT ISZERO PUSH2 0x147 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST DUP1 CALLDATALOAD SWAP1 PUSH1 0x20 ADD SWAP2 DUP5 PUSH1 0x1 DUP4 MUL DUP5 ADD GT PUSH5 0x100000000 DUP4 GT OR ISZERO PUSH2 0x169 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST SWAP2 SWAP1 DUP1 DUP1 PUSH1 0x1F ADD PUSH1 0x20 DUP1 SWAP2 DIV MUL PUSH1 0x20 ADD PUSH1 0x40 MLOAD SWAP1 DUP2 ADD PUSH1 0x40 MSTORE DUP1 SWAP4 SWAP3 SWAP2 SWAP1 DUP2 DUP2 MSTORE PUSH1 0x20 ADD DUP4 DUP4 DUP1 DUP3 DUP5 CALLDATACOPY PUSH1 0x0 DUP2 DUP5 ADD MSTORE PUSH1 0x1F NOT PUSH1 0x1F DUP3 ADD AND SWAP1 POP DUP1 DUP4 ADD SWAP3 POP POP POP POP POP POP POP SWAP2 SWAP3 SWAP2 SWAP3 SWAP1 POP POP POP PUSH2 0x22B JUMP JUMPDEST PUSH1 0x40 MLOAD DUP1 DUP3 PUSH28 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF NOT AND PUSH28 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF NOT AND DUP2 MSTORE PUSH1 0x20 ADD SWAP2 POP POP PUSH1 0x40 MLOAD DUP1 SWAP2 SUB SWAP1 RETURN JUMPDEST DUP1 PUSH1 0x0 DUP1 PUSH2 0x100 EXP DUP2 SLOAD DUP2 PUSH1 0xFF MUL NOT AND SWAP1 DUP4 ISZERO ISZERO MUL OR SWAP1 SSTORE POP POP JUMP JUMPDEST PUSH1 0x0 DUP1 PUSH1 0x0 SWAP1 SLOAD SWAP1 PUSH2 0x100 EXP SWAP1 DIV PUSH1 0xFF AND ISZERO PUSH2 0x250 JUMPI PUSH4 0x20C13B0B PUSH1 0xE0 SHL SWAP1 POP PUSH2 0x258 JUMP JUMPDEST PUSH1 0x0 PUSH1 0xE0 SHL SWAP1 POP JUMPDEST SWAP3 SWAP2 POP POP JUMP INVALID LOG2 PUSH6 0x627A7A723158 KECCAK256 SWAP14 PUSH27 0xA06B7443AA12CEE8B1BA4356AF624DE6B912D6AF47B494C9B9D621 0xb8 DUP4 0xac PUSH5 0x736F6C6343 STOP SDIV SIGNEXTEND STOP ORIGIN ",
  sourceMap:
    "0:350:0:-;;;46:5;31:20;;;;;;;;;;;;;;;;;;;;0:350;8:9:-1;5:2;;;30:1;27;20:12;5:2;0:350:0;;;;;;;",
};

const GANACHE_CONF = {
  seed: "0xd30553e27ba2954e3736dae1342f5495798d4f54012787172048582566938f6f",
};
const GANACHE_CHAIN_ID = 1337;
const send = (provider: any, data: any): Promise<any> =>
  new Promise((resolve, reject) =>
    provider.send(data, (err: any, res: any) => {
      if (err) reject(err);
      else resolve(res.result);
    })
  );

let provider: any, addresses: string[], contractAddress: string;

const testDid = "did:3:bafysdfwefwe";

beforeAll(async () => {
  provider = ganache.provider(GANACHE_CONF);
  addresses = await send(provider, linking.encodeRpcMessage("eth_accounts"));
  // ganache-core doesn't support personal_sign -.-
  provider.manager.personal_sign = (
    data: any,
    address: string,
    callback: any
  ): void => {
    // next line is hack to make contract address to personal sign
    if (address === contractAddress.toLowerCase()) address = addresses[0];
    const account = provider.manager.state.accounts[address.toLowerCase()];
    const result = sigUtils.personalSign(account.secretKey, { data });
    callback(null, result);
  };
  // deploy contract wallet
  const factory = new ContractFactory(
    CONTRACT_WALLET_ABI,
    CONTRACT_WALLET_BYTECODE
  );
  const unsignedTx = Object.assign(factory.getDeployTransaction(), {
    from: addresses[0],
    gas: 4712388,
    gasPrice: 100000000000,
    nonce: 0,
  });
  await send(provider, linking.encodeRpcMessage("eth_sendTransaction", [unsignedTx]));
  contractAddress = Contract.getContractAddress(unsignedTx);
  // mock ethers providers
  (providers as any).getNetwork = (): any => {
    return {
      _defaultProvider: (): any => {
        return new providers.Web3Provider(provider);
      },
    };
  };
});

test("invalid ethereumEOA proof should return null", async () => {
  // wrong address
  const account = new AccountID({ address: addresses[1], chainId: "eip155:1" });
  const invalidProof = { account } as unknown as LinkProof;
  expect(await ethereum.validateLink(invalidProof)).toBeFalsy();
  // invalid signature
  const invalidSignatureProof = Object.assign(
    {},
    {
      signature:
        "0xfa69ccf4a94db6132542abcabcabcab234b73f439700fbb748209890a5780f3365a5335f82d424d7f9a63ee41b637c116e64ef2f32c761bb065e4409f978c4babc",
    }
  ) as unknown as LinkProof;
  expect(await ethereum.validateLink(invalidSignatureProof)).toBeFalsy();
});

test("validateLink: valid ethereumEOA proof should return proof", async () => {
  const authProvider = new linking.ethereum.EthereumAuthProvider(
    provider,
    addresses[0]
  );
  const proof = await authProvider.createLink(testDid);
  await expect(ethereum.validateLink(proof)).resolves.toEqual(proof);
});

test("validate v0 and v1 proofs", async () => {
  expect(await ethereum.validateLink(proofs.v0.valid as unknown as LinkProof)).toMatchSnapshot();
  await expect(ethereum.validateLink(proofs.v0.invalid as unknown as LinkProof)).resolves.toEqual(null);
  expect(await ethereum.validateLink(proofs.v1.valid as unknown as LinkProof)).toMatchSnapshot();
  expect(await ethereum.validateLink(proofs.v1.invalid as unknown as LinkProof)).toEqual(null);
});

test("invalid erc1271 proof should return null", async () => {
  // the contract wallet we deployed should just return false by default
  // when trying to validate signature
  const account = new AccountID({
    address: contractAddress,
    chainId: "eip155:" + GANACHE_CHAIN_ID,
  });
  const erc1271Proof = { account, type: "erc1271" } as unknown as LinkProof
  expect(await ethereum.validateLink(erc1271Proof)).toBeFalsy();
});

test("validateLink: valid erc1271 proof should return proof", async () => {
  // tell the contract wallet contract to return valid signature instead
  const contract = new Contract(
    contractAddress,
    CONTRACT_WALLET_ABI,
    new providers.Web3Provider(provider)
  );
  let tx = await contract.populateTransaction.setIsValid(true);
  tx = Object.assign(tx, {
    from: addresses[0],
    gas: 4712388,
    gasPrice: 100000000000,
  });
  await send(provider, linking.encodeRpcMessage("eth_sendTransaction", [tx]));
  const authProvider = new linking.ethereum.EthereumAuthProvider(
    provider,
    contractAddress
  );
  const proof = await authProvider.createLink('msg')
  expect(await ethereum.validateLink(proof)).toEqual(proof);
});
