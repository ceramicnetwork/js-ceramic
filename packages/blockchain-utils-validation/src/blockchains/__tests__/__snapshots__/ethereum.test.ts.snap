// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate v0 and v1 proofs 1`] = `
Object {
  "account": "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1",
  "message": "Create a new 3Box profile

- 
Your unique profile ID is did:3:bafysdfwefwe",
  "signature": "0xfa69ccf4a94db61325429d37c58c6de534b73f439700fbb748209890a5780f3365a5335f82d424d7f9a63ee41b637c116e64ef2f32c761bb065e4409f978c4bb1c",
  "type": "ethereum-eoa",
  "version": 2,
}
`;

exports[`validate v0 and v1 proofs 2`] = `
Object {
  "account": "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1",
  "message": "Create a new 3Box profile

- 
Your unique profile ID is did:3:bafysdfwefwe",
  "signature": "0xfa69ccf4a94db61325429d37c58c6de534b73f439700fbb748209890a5780f3365a5335f82d424d7f9a63ee41b637c116e64ef2f32c761bb065e4409f978c4bb1c",
  "type": "ethereum-eoa",
  "version": 2,
}
`;
